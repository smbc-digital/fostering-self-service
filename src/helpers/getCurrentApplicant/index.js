import { Applicant, SECOND_APPLICANT_ROUTE } from 'constants'

const getCurrentApplicant = match => {
    if (match.params && match.params[0] === SECOND_APPLICANT_ROUTE) {
        return Applicant.SecondApplicant
    }

    return Applicant.FirstApplicant
}

export default getCurrentApplicant