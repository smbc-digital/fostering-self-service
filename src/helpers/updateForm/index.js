import { fetchWithTimeout } from '../index'

const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

const StageName = {
    HomeVisit: 0,
    Application: 1
}

const HomeVisitFormName = {
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguagesSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 8
}

const ApplicationFormName = {
    References: 0,
    GpDetails: 1,
    AddressHistory: 2
}

const getHomeVisitUpdateEndpoint = form => {
    switch (form) {
        case HomeVisitFormName.TellUsAboutYourself:
            return '/fostering/home-visit/about-yourself'
        case HomeVisitFormName.YourEmploymentDetails:
            return '/fostering/home-visit/your-employment-details'
        case HomeVisitFormName.LanguagesSpokenInYourHome:
            return '/fostering/home-visit/languages-spoken-in-your-home'
        case HomeVisitFormName.YourFosteringHistory:
            return '/fostering/home-visit/your-fostering-history'
        case HomeVisitFormName.YourPartnership:
            return '/fostering/home-visit/partnership-status'
        case HomeVisitFormName.TellUsAboutYourInterestInFostering:
            return '/fostering/home-visit/interest-in-fostering'
        case HomeVisitFormName.YourHealth:
            return '/fostering/home-visit/about-your-health'
        case HomeVisitFormName.YourHousehold:
            return '/fostering/home-visit/household'
        case HomeVisitFormName.ChildrenLivingAwayFromYourHome:
            return '/fostering/home-visit/children-living-away-from-home'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getApplicationUpdateEndpoint = form => {
    switch (form) {
        case ApplicationFormName.References:
            return '/fostering/application/references'
        case ApplicationFormName.GpDetails:
            return '/fostering/application/gp-details'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getFormStageEndpoint = stage => {
    switch (stage) {
        case StageName.HomeVisit: 
            return '/fostering/home-visit'
        case StageName.Application:
            return '/fostering/application'
        default:
            throw new Error('No matching endpoint for given stage.')
    }
}

const reduceProperties = object => Object.keys(object).reduce((acc, property) => {
    return {
        ...acc,
        [property]: object[property].value
    }
}, {})

const updateForm = async (endpoint, formData) => {
    const parsedFormData = parseFormData(formData)

    const response = await fetchWithTimeout(endpoint, {
        method: 'PATCH',
        credentials: 'include',
        body: JSON.stringify(parsedFormData),
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
    }, 30000)

    if (!response.ok) {
        throw Error(response.statusText)
    }

    return await response.json()
}

const updateFormStatus = ({ form, stage, currentStatus, setStatus }) => {
    if (currentStatus === TaskStatus.None) {
        fetchWithTimeout(`${getFormStageEndpoint(stage)}/status`,
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                }),
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            }, 30000)

        setStatus(TaskStatus.NotCompleted)
    }
}

const parseFormData = ({ firstApplicant, secondApplicant, familyReference, firstPersonalReference, secondPersonalReference, ...formData }) => {
    let parsedObject = reduceProperties(formData)

    if (firstApplicant) {
        parsedObject.firstApplicant = reduceProperties(firstApplicant)
    }

    if (secondApplicant) {
        parsedObject.secondApplicant = reduceProperties(secondApplicant)
    }

    if (familyReference) {
        parsedObject.familyReference = reduceProperties(familyReference)
        parsedObject.familyReference.address = familyReference.address.value
    }

    if (firstPersonalReference) {
        parsedObject.firstPersonalReference = reduceProperties(firstPersonalReference)
        parsedObject.firstPersonalReference.address = firstPersonalReference.address.value
    }

    if (secondPersonalReference) {
        parsedObject.secondPersonalReference = reduceProperties(secondPersonalReference)
        parsedObject.secondPersonalReference.address = secondPersonalReference.address.value
    }

    return parsedObject
}

const updateApplicationForm = async (form, formData) => {
    const endpoint = getApplicationUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

const updateHomeVisitForm = async (form, formData) => {
    const endpoint = getHomeVisitUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

export {
    TaskStatus,
    StageName,
    HomeVisitFormName,
    ApplicationFormName,
    updateFormStatus,
    parseFormData,
    updateApplicationForm,
    updateHomeVisitForm
}