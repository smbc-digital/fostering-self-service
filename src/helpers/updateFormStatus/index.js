export const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

export const FormName =
{
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguageSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 7
}

export default function (form, currentStatus, setStatus) {
    if (currentStatus === TaskStatus.None) {
        fetch('/front-end-endpoint-update-status',
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                })
            })

        setStatus(TaskStatus.NotCompleted)
    }
}