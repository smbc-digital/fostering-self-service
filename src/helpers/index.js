import fetchWithTimeout from './fetchWithTimeout'
import { getPageRoute } from './pagehelper'
import getCurrentApplicant from './getCurrentApplicant'
import { updateFormStatus, updateForm, FormName, TaskStatus } from './updateForm'

export {
    fetchWithTimeout,
    getPageRoute,
    getCurrentApplicant,
    updateFormStatus,
    updateForm,
    FormName,
    TaskStatus
}