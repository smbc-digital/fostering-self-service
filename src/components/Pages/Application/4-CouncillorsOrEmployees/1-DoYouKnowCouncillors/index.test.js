import { React, mount, useContextMock, renderer } from 'helpers/SetupTest'
import { act } from 'react-dom/test-utils'
import DoYouKnowCouncillors from './index'
import * as helpers from 'helpers'
import { 
    ABOUT_RELATIONSHIP_TO_COUNCIL_EMPLOYEES, 
    START_PAGE,
    RELATIONSHIP_TO_COUNCIL_EMPLOYEES 
} from 'routes'

describe('DoYouKnowCouncillors', () => {

    const onChangeStatusMock = jest.fn()
    const onChangeTargetMock = jest.fn()

    beforeEach(() => {
        useContextMock.mockReturnValue({
            firstApplicant: {
                hasContactWithCouncillor: {
                    value: '',
                    isValid: false
                }
            },
            onChangeTarget: onChangeTargetMock,
            onChangeStatus: onChangeStatusMock,
            statuses: {
                councillorsOrEmployeesStatus: 0
            }
        })
    })

    it('should call updateApplicationForm on submit', async () => {
        // Arrange
        useContextMock.mockReturnValue({
            firstApplicant: {
                hasContactWithCouncillor: {
                    value: 'false',
                    isValid: false
                }
            },
            onChangeTarget: jest.fn(),
            statuses: {
                councillorsOrEmployeesStatus: 0
            },
            onChangeStatus: onChangeStatusMock
        })
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }
        helpers.updateApplicationForm = jest.fn().mockReturnValue(Promise.resolve(0)) 

        // Act
        const wrapper = mount(<DoYouKnowCouncillors history={history} match={match}/>)

        await act(async () => {
            await wrapper.find('form').simulate('submit')
            await Promise.resolve()
        })

        // Assert
        expect(helpers.updateApplicationForm).toHaveBeenCalled()
    })
    
    const nextPageTestInputs = [
        {
            expectedRoute: ABOUT_RELATIONSHIP_TO_COUNCIL_EMPLOYEES,
            matchParams: [],
            firstApplicant: {
                hasContactWithCouncillor: {
                    value: true,
                    isValid: false
                },
                firstName: {
                    value: 'first name'
                },
                lastName: {
                    value: 'last name'
                }
            },
            secondApplicant: {}
        },
        {
            expectedRoute: `${ABOUT_RELATIONSHIP_TO_COUNCIL_EMPLOYEES}/second-applicant`,
            matchParams: ['second-applicant'],
            firstApplicant: {},
            secondApplicant: {
                hasContactWithCouncillor: {
                    value: true,
                    isValid: false
                },
                firstName: {
                    value: 'first name'
                },
                lastName: {
                    value: 'last name'
                }
            }
        },
        {
            expectedRoute: START_PAGE,
            matchParams: [],
            firstApplicant: {
                hasContactWithCouncillor: {
                    value: false,
                    isValid: false
                }
            }
        },
        {
            expectedRoute: `${RELATIONSHIP_TO_COUNCIL_EMPLOYEES}/second-applicant`,
            matchParams: [],
            firstApplicant: {
                hasContactWithCouncillor: {
                    value: false,
                    isValid: false
                },
                firstName: {
                    value: 'first name'
                },
                lastName: {
                    value: 'last name'
                }
            },
            secondApplicant: {}
        }
    ]

    nextPageTestInputs.forEach(({ expectedRoute, firstApplicant, matchParams, secondApplicant }) => {
        it('should push to the next page on submit', async () => { 
            // Arrange
            useContextMock.mockReturnValue({
                firstApplicant,
                secondApplicant, 
                onChangeTarget: jest.fn(),
                onChangeStatus: jest.fn(),
                statuses: {
                    councillorsOrEmployeesStatus: 0
                }
            })
            helpers.updateApplicationForm = jest.fn().mockReturnValue(Promise.resolve(0)) 
            const history = {
                push: jest.fn()
            }
            const match = {
                params: matchParams
            }

            // Act
            const wrapper = mount(<DoYouKnowCouncillors history={history} match={match} />)
            await act(async () => {
                await wrapper.find('form').simulate('submit')
                await Promise.resolve()
            })

            // Assert
            expect(history.push).toHaveBeenCalledWith(expectedRoute)
        })
    })

    it('should set isLoading', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }

        // Make the update throw an error since isLoading is set back to false on success
        helpers.updateApplicationForm = jest.fn().mockImplementation(() => { throw new Error() })

        // Act
        const wrapper = mount(<DoYouKnowCouncillors history={history} match={match} />)
        await act(async () => {
            await wrapper.find('form').simulate('submit')
            wrapper.update()
        })
        
        // Assert
        expect(wrapper.find('Button').first().props().isLoading).toEqual(true)
    })

    it('should update form status', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }
        helpers.updateApplicationForm = jest.fn().mockReturnValue(Promise.resolve(0)) 

        // Act
        const wrapper = mount(<DoYouKnowCouncillors history={history} match={match}/>)
        await act(async () => {
            await wrapper.find('form').simulate('submit')
            await Promise.resolve()
        })

        // Assert
        expect(onChangeStatusMock).toHaveBeenCalledWith('councillorsOrEmployeesStatus', 0)
    })

    it('should push to error page on submit', () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }
        helpers.updateApplicationForm = jest.fn().mockImplementation(() => { throw new Error() })

        // Act
        const wrapper = mount(<DoYouKnowCouncillors history={history} match={match}/>)
        wrapper.find('form').simulate('submit')

        // Assert
        expect(history.push).toHaveBeenCalledWith('/error')
    })

    it('should push to start page on save and go back click', async () => {
        // Arrange
        useContextMock.mockReturnValue({
            firstApplicant: {
                hasContactWithCouncillor: {
                    value: false,
                    isValid: true
                }
            },
            statuses: {
                councillorsOrEmployeesStatus: 0
            },
            onChangeTarget: jest.fn(),
            onChangeStatus: onChangeStatusMock
        })
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }
        helpers.updateApplicationForm = jest.fn().mockReturnValue(Promise.resolve(0)) 

        // Act
        const wrapper = mount(<DoYouKnowCouncillors history={history} match={match}/>)
        await act(async () => {
            await wrapper.find('button').at(1).simulate('click')
            await Promise.resolve()
        })

        // Assert
        expect(history.push).toHaveBeenCalledWith(START_PAGE)
    })

    it('should call updateFormStatus', () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        const match = {
            params: []
        }
        helpers.updateFormStatus = jest.fn() 

        // Act
        mount(<DoYouKnowCouncillors history={history} match={match} />)

        // Assert
        expect(helpers.updateFormStatus).toHaveBeenCalled()
    })

    it('should call onChangeTarget', () => {
        // Arrange 
        const history = {
            push: jest.fn()
        }

        const match = {
            params: []
        }

        // Act
        const wrapper = mount(<DoYouKnowCouncillors history={history} match={match} />)
        wrapper.find('input').first().simulate('change')

        // Assert
        expect(onChangeTargetMock).toHaveBeenCalled()
    })

    describe('snapshot', () => {
        it('should render correctly', () => {
            const tree = renderer
                .create(<DoYouKnowCouncillors history={{}} match={{ params: [] }} />)
                .toJSON()

            expect(tree).toMatchSnapshot()
        })
    })
})