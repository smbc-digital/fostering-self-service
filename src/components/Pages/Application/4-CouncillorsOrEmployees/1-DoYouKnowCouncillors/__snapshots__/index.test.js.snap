// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoYouKnowCouncillors snapshot should render correctly 1`] = `
Array [
  <h1>
    Your fostering journey
  </h1>,
  <h2>
    Personal relationships with local councillors or council employees
  </h2>,
  <form
    onSubmit={[Function]}
  >
    <h3
      className="no-margin-bottom"
    >
      Do you have a personal relationship with a local councillor or council employee?
    </h3>
    <fieldset>
      <div
        className="input-radio-container"
        role="radiogroup"
      >
        <input
          checked={false}
          disabled={undefined}
          id="has-contact-with-councillor-yes"
          name="hasContactWithCouncillor"
          onChange={[Function]}
          type="radio"
          value="true"
        />
        <label
          htmlFor="has-contact-with-councillor-yes"
        >
          Yes
        </label>
      </div>
      <div
        className="input-radio-container"
        role="radiogroup"
      >
        <input
          checked={false}
          disabled={undefined}
          id="has-contact-with-councillor-no"
          name="hasContactWithCouncillor"
          onChange={[Function]}
          type="radio"
          value="false"
        />
        <label
          htmlFor="has-contact-with-councillor-no"
        >
          No
        </label>
      </div>
    </fieldset>
    <button
      className="button-primary"
      disabled={true}
      id={undefined}
      onClick={undefined}
      type="submit"
    >
      Next step
    </button>
    <p>
      <a
        href="#"
        onClick={[Function]}
      >
        Previous
      </a>
    </p>
    <button
      className="button-disabled left-chevron"
      disabled={true}
      id={undefined}
      onClick={[Function]}
      type="submit"
    >
      Save and go back to fostering area
    </button>
  </form>,
]
`;
